src/commands/gw2.ts

```diff
@@ -0,0 +1,165 @@
+        "API key required. Please check your app-config."
+    );
+    return undefined;
+  }
+
+   const commands = {
+       title: buildTitleCommand(),
+       characters: buildCharactersCommand(gw2APIKey),
+       wallet: buildCurrencyCommand(gw2APIKey),
+    };
+  return buildNestedCommand(prefix, "gw2", "Surface GW2 information.", commands);
+}
+
+function buildTitleCommand() {
+    return {
+        run: (message, parameters) => {
+
+          if (parameters.length !== 1) {
+            message.reply("Expecting one title id");
+            return;
+          }
+
+          if (isNaN(Number(parameters[0]))) {
+            message.reply("Invalid title id expected number");
+            return;
+          }
+
+            getTitle(Number(parameters[0]))
+            .then((text) => message.channel.send(text[0]["name"]))
+            .catch((err) => message.reply(err));
+        },
+        help: "Show title name given title id",
+    };
+}
+
+function buildCharactersCommand(gw2APIKey){
+    return {
+        run: (message) => {
+            getCharacters(gw2APIKey)
+            .then((text) => message.channel.send(text))
+            .catch((err) => message.reply(err));
+        },
+        help: "Show character names",
+    };
+}
+
+function buildCurrencyCommand(gw2APIKey){
+  return {
+      run: (message, parameters) => {
+
+        const currencyName = parameters.join(" ");
+        if (parameters.length === 0) {
+          message.reply("Expecting currency name");
+          return;
+        }
+
+        getCurrencies()
+        .then( (all) => {
+            const currency = all.find(walletItem => walletItem["name"] == currencyName);
+            if (!currency){
+              message.reply("Invalid currency name");
+              return;
+            }
+            const currencyid = currency["id"];
+            getWallet(gw2APIKey)
+            .then( (wallet) => {
+              const currencyInWallet = wallet.find(walletItem => walletItem["id"] === currencyid);
+              const currencyAmount = currencyInWallet["value"];
+              return currencyAmount;
+            })
+            .then((currencyAmount) => message.channel.send("You have " + currencyAmount + " " + currencyName + "/s"))
+            .catch((err) => message.reply(err));
+        });
+      },
+      help: "Show how much of a currency u have",
+  };
+}
+
+/**
+ * Get GW2 titles.
+ */
+async function getTitle(titleID) {
+    const url = `https://api.guildwars2.com/v2/titles?ids=${titleID}`;
+      return fetch(url).then((res) => {
+      if (res.ok) {
+        return res.json();
+      }
+  
+      console.error(`Unable to fetch titles: ${res.status} ${res.statusText}`);
+      throw "Unable to fetch titles info. Please contact the bot maintainer if issues persist.";
+    });
+  }
+
+/**
+ * Get GW2 characters.
+ */
+async function getCharacters(gw2APIKey) {
+    const url = `https://api.guildwars2.com/v2/characters?access_token=${gw2APIKey}`;
```

Comment:
You mentioned the issue earlier and I think you're right: commands that need the API key will only really be useful if we can individually register our personal API keys.

There are 2 ways that pop up in my head:

1. We hardcode API keys for each of us and according to which Discord user-id makes the call we then use the appropriate key
1. We add a command that can register API keys on the fly. This'd need integration with a database. There's already a central database setup for the `events` command. I think we can hook this functionality into that database too. The main headache here is that I haven't set up a good dev-story for working with a database: I was playing with the actual table when I set up event command.

What do you think?